all_seaing_driver/all_seaing_driver/thrust_commander.py

```diff
@@ -47,12 +61,74 @@ def back_right_cb(self, msg: Int64):
 
     def back_left_cb(self, msg: Int64):
         self.send_pwm(self.back_left_port, msg.data)
+    
+    def get_time(self):
+        """
+        Get the current time. Written into a function because it's annoying to write every time.
+        """
+        return self.get_clock().now()
+    
+    def turn_off_thrusters(self):
+        """
+        Turn off all thrusters.
+        """
+        self.send_pwm(self.front_right_port, 1500)
+        self.send_pwm(self.front_left_port, 1500)
+        self.send_pwm(self.back_right_port, 1500)
+        self.send_pwm(self.back_left_port, 1500)
 
     def send_pwm(self, channel, value):
         self.get_logger().info(f"Sending PWM value {value} to channel {channel}")
         return self.proxy.call_async(
             CommandLong.Request(command=183, param1=float(channel), param2=float(value))
         )
+    
+    def get_thrust_values(self, tx, ty, tn):
+        """
+        Given linear and angular velocities, get thrust velocities for each thruster.
+        """
+        d = 2 ** (3 / 2)
+        ang = tn / (4 * self.r)
+        return {
+            "back_right": (tx - ty) / d + ang,
+            "back_left": (ty + tx) / d - ang,
+            "front_left": (tx - ty) / d - ang,
+            "front_right": (ty + tx) / d + ang,
+        }
+    
+    def receive_heartbeat(self, msg):
+        """
+        Callback function for receiving hearbeat messages, which are necessary for the controller to run.
+        """
+        self.get_logger().info("Heartbeat received!")
+        if msg.e_stopped: 
+            self.turn_off_thrusters()
+            raise Exception("Received e-stop message! Killing node")
+
+        self.last_heartbeat_timestamp = self.get_time()
+
+    def restrict_input(self, input, max_val):
+        if input < -max_val:
+            return -max_val
+        if input > max_val:
+            return max_val
+        return input
+    
+    def timer_callback(self):
+        current_time = self.get_time()
+        time_since_last_heartbeat = (current_time - self.last_heartbeat_timestamp).nanoseconds / 1e9
+
+        if time_since_last_heartbeat > self.required_heartbeat_recentness:
+            # Heartbeat timeout - stop thrusters
+            self.turn_off_thrusters()
+            raise Exception("Lost heartbeat! Killing node")
```

Comment:
Instead of killing the node, let's do something like (if estop received, then we go into estopped mode where we constantly publish zero commands (`self.turn_off_thrusters`) in the timer callback AND the callbacks don't call the send_pwm function anymore.

Reply:
ok i added this logic and the other changes (i tested the loosing heartbeat stuff on rviz and it works i think?) lmk if i should change anything else